readme.md
# ![HawtDispatch][logo]
[logo]: http://hawtdispatch.fusesource.org/images/project-logo.png

## Synopsis

[HawtDispatch][] is a small ( less than 100k ) thread pooling and NIO
handling framework API modeled after the `libdispatch` API. `libdispatch` is
the API that Apple created to power the Grand Central Dispatch (GCD)
technology in OS X. It allows you to more easily develop multi-threaded
applications which can more easily scale to take advantage of all the
processing cores on your machine. At the same time, it's development model
simplifies solving many of the problems that plague multi-threaded NIO
development.

[HawtDispatch]:http://hawtdispatch.fusesource.org

## Resources

* [Developer Guide](http://hawtdispatch.fusesource.org)
* [API Reference](http://hawtdispatch.fusesource.org/maven/1.0/hawtdispatch/apidocs/org/fusesource/hawtdispatch/package-summary.html)

## Building from Source

Prerequisites:

* [Maven >= 2.2.1](http://maven.apache.org/download.html)
* [Java JDK >= 1.6](http://java.sun.com/javase/downloads/widget/jdk6.jsp)

In the this directory, run:

    mvn install



changelog.md
# ![HawtDispatch](http://hawtdispatch.fusesource.org/images/project-logo.png)
=============================================================================

## [HawtDispatch 1.19](http://hawtdispatch.fusesource.org/blog/releases/release-1-19.html),  released 2014-01-23

* You can now configure the what cypher suites are enabled on a SSLTransport via the enabledCypherSuites property
* add gradle build script
* Fix range check on timeout, as it always evaluated to true
* Fixing bugs where the read/write rate limiting features of the tcp transport might not take effect.

## [HawtDispatch 1.18](http://hawtdispatch.fusesource.org/blog/releases/release-1-18.html),  released 2013-10-11

* Do not even invoke getLocalHost() if not necessary
* propagate name resolution exceptions in the correct thread context
* Fixes issue #11 : Typo in TcpTransportServer

## [HawtDispatch 1.17](http://hawtdispatch.fusesource.org/blog/releases/release-1-17.html),  released 2013-05-18

* Add a get method to Future objects.
* Fixing bug where SSL transports were not getting past the SSL handshake (on linux).

## [HawtDispatch 1.16](http://hawtdispatch.fusesource.org/blog/releases/release-1-16.html), released 2013-05-10

* Fixing invalid the metric duration value.
* Do a better job ignoring errors when JMX is not supported by the platform. (Android etc.)
* Fixes issue #9: HawtDispatch does not work on Android 4.0 

## [HawtDispatch 1.15](http://hawtdispatch.fusesource.org/blog/releases/release-1-15.html), released 2013-04-19

* Also include info about how long the metric was monitored for.
* Expose dispatch queue profiling data via JMX.

## [HawtDispatch 1.14](http://hawtdispatch.fusesource.org/blog/releases/release-1-14.html), released 2013-04-09

* Fixes bug where you could end up in a CPU spin when the SSL session closed.
* Added a disabledCypherSuites property to the SSL transports to support disabling some of the supported cypher suites.
* Add a getThread() method to the ThreadDispatchQueue class.

## [HawtDispatch 1.13](http://hawtdispatch.fusesource.org/blog/releases/release-1-13.html), released 2012-12-23

* Upgrade to scala 2.10
* Avoid possible NPE.
* Setup the client SSL transports so that the internal session reuse strategy can be used.

## [HawtDispatch 1.12](http://hawtdispatch.fusesource.org/blog/releases/release-1-12.html), released 2012-09-20

* Custom dispatch sources will now return null after the event has been received to avoid accidentally double processing events.
* Make sure we only return false from the offer method when the transport is also full()
* Add a closeOnCancel option to disable closing the socket when the transport is stopped.
* Rename the SecuredTransport interface to SecuredSession and now both the SSLProtocolCodec and SslTransport implement it.
* You can now handle SSL/TLS encoding/decoding via a wrapping protocol codec.  ProtocolCodec and Transport interfaces needs a couple of adjustments to properly support the co
* Better handling of getting the local host address.
* Protocol codec decoding optimizations.
* Move all the connect logic into the start method.
* Do host name resolutions on a blocking executor to avoid stalling the hawtdispatch threads.
* Resize the read buffer after reading from the channel if to avoid to avoid holding on to large buffers.
* Support changing the socket send/recv buffer size on started transports and servers.
* Do at least a non blocking select when we notice that another thread requested the NIO manager wakeup.  This should allow us to pickup any new IO events that have occurred 
* Dropped the continuations example, added a SSL transport client example

## [HawtDispatch 1.11](http://hawtdispatch.fusesource.org/blog/releases/release-1-11.html), released 2012-05-02

* Support buffer pooling in the abstract protocol codec.
* Adding a TransportAware interface which codecs can implement to be injected with their associated codec.
* Make it easy to configure the client auth mode on the ssl transport server.
* Fixes SSL transport bug where inbound data was not delivered after SSL handshake completed.
* Allow a SslTransportServer to be created just using the constructor.

## [HawtDispatch 1.10](http://hawtdispatch.fusesource.org/blog/releases/release-1-10.html), released 2012-04-06

* Fix assertion error message.
* Switch to using 'Task' abstract base instead of the Runnable interface, in many cases it improves perf by 20% since virtual invocations are cheaper then interface invocation
* Adding a UDP based transport.
* Support configuring the client auth mode.

## [HawtDispatch 1.9](http://hawtdispatch.fusesource.org/blog/releases/release-1-9.html), released 2012-02-27

* Fixes LOW priority global queue was being created with a high thread priority.
* Distribute work load spikes fairly to the worker threads.
* Support updating the transport's dispatch queue.
* Add assertion that catches the following error: queue.setTargetQueue(queue)
* Adding a SecureTransport interface and removing many of the leaky abstractions in the transport package.

## [HawtDispatch 1.8](http://hawtdispatch.fusesource.org/blog/releases/release-1-8.html), released 2012-01-30

* Fixes CPU spin that occurred when a peer disconnected while SSL handshake is in progress.
* Could not create a client initiated SSL connection.

## [HawtDispatch 1.7](http://hawtdispatch.fusesource.org/blog/releases/release-1-7.html), released 2012-01-13

* Cleaning up the transport interfaces. Added an abstract protocol codec that makes it easier to implement them.

## [HawtDispatch 1.6](http://hawtdispatch.fusesource.org/blog/releases/release-1-6.html), released 2011-12-19

* Support using a system property to configure the number of dispatch threads.
* Added a hawtdispatch-transport module which provides nice Transport abstraction for working with Sockets and HawtDispatch.

## [HawtDispatch 1.5](http://hawtdispatch.fusesource.org/blog/releases/release-1-5.html), released 2011-11-29


hawtdispatch-website/ext/Website.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import org.fusesource.scalate.RenderContext

package

/**
 * <p>
 * </p>
 *
 * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
 */
object Website {

  val project_name= "HawtDispatch"
  val project_slogan= "The libdispatch style API for Java and Scala"
  val project_id= "hawtdispatch"
  val project_jira_key= "HAWTDISPATCH"
  val project_issue_url= "http://hawtdispatch.assembla.com/spaces/hawtdispatch/support/tickets"
  val project_forums_url= "http://groups.google.com/group/hawtdispatch"
  val project_wiki_url= "http://wiki.github.com/fusesource/hawtdispatch"
  val project_logo= "/images/project-logo.png"
  val project_version= "1.16"
  val project_snapshot_version= "1.17-SNAPSHOT"
  val project_versions = List(
        project_version,
        "1.15",
        "1.14","1.13","1.12","1.11","1.10",
        "1.9","1.8","1.7","1.6","1.5","1.4","1.3","1.2","1.1","1.0")  

  val project_keywords= "dispatch,executor,java,scala,libdispatch,gcd,actor,thread,pool"

  // -------------------------------------------------------------------
  val github_page = "http://github.com/fusesource/hawtdispatch"
  val git_user_url = "git://github.com/fusesource/hawtdispatch.git"
  val git_commiter_url = "git@github.com:fusesource/hawtdispatch.git"
  
  val project_maven_groupId= "org.fusesource.hawtdispatch"
  val project_maven_artifactId= "hawtdispatch"

  val website_base_url= "http://hawtdispatch.fusesource.org"
  
  val api_dir = "http://hawtdispatch.fusesource.org/versions/"+project_version+"/maven/hawtdispatch/apidocs/org/fusesource/hawtdispatch"
  val scala_api_dir = "http://hawtdispatch.fusesource.org/versions/"+project_version+"/maven/hawtdispatch-scala/scaladocs"

}